src/buildifier/buildifier.ts

```diff
@@ -0,0 +1,211 @@
+  // automatic format detection (see:
+  // https://github.com/bazelbuild/buildtools/blob/d39e4d/build/lex.go#L88)
+  // so that user actions in the IDE are consistent with the behavior they
+  // would see running buildifier on the command line.
+  const parsedPath = path.parse(fsPath.toLowerCase());
+  if (parsedPath.ext === ".bzl" || parsedPath.ext === ".sky") {
+    return "bzl";
+  }
+  if (parsedPath.ext === ".build" || parsedPath.base === "build") {
+    return "build";
+  }
+  if (parsedPath.ext === ".workspace" || parsedPath.base === "workspace") {
+    return "workspace";
+  }
+  return "bzl";
+}
+
+/**
+ * Gets the path to the buildifier executable specified by the workspace
+ * configuration, if present.
+ *
+ * @returns The path to the buildifier executable specified in the workspace
+ *     configuration, or just "buildifier" if not present (in which case the
+ *     system path will be searched).
+ */
+export function getDefaultBuildifierExecutablePath(): string {
+  // Try to retrieve the executable from VS Code's settings. If it's not set,
+  // just use "buildifier" as the default and get it from the system PATH.
+  const bazelConfig = vscode.workspace.getConfiguration("bazel");
+  const buildifierExecutable = bazelConfig.buildifierExecutable as string;
+  if (buildifierExecutable.length === 0) {
+    return "buildifier";
+  }
+  return buildifierExecutable;
+}
+
+/**
+ * Parses the output of buildifier's {@code --lint=warn} mode and constructs
+ * objects representing the warnings.
+ *
+ * @param lines The lines of output from standard error.
+ */
+function parseLintWarnings(lines: string[]): BuildifierWarning[] {
+  const warnings = new Array<BuildifierWarning>();
+  let lineNumber = 0;
+  let category = "";
+  let message = "";
+
+  for (const line of lines) {
+    if (line.startsWith("stdin:")) {
+      if (message) {
+        warnings.push(new BuildifierWarning(lineNumber, category, message));
+      }
+      const [_, lineNumberPart, categoryPart, ...remainder] = line.split(":");
+      lineNumber = parseInt(lineNumberPart, 10);
+      category = categoryPart.trim();
+      message = remainder.join(":");
+    } else {
+      message += `\n${line}`;
+    }
+  }
+
+  if (message) {
+    warnings.push(new BuildifierWarning(lineNumber, category, message));
+  }
+
+  return warnings;
+}
+
+/**
+ * Executes buildifier with the given file content and arguments.
+ *
+ * @param fileContent The BUILD or .bzl file content to process, which is sent
+ *     via stdin.
+ * @param args Command line arguments to pass to buildifier.
+ */
+function executeBuildifier(
+  fileContent: string,
+  args: string[],
+): Promise<{ stdout: string; stderr: string }> {
+  return new Promise((resolve, reject) => {
+    const execOptions = {
+      maxBuffer: Number.MAX_SAFE_INTEGER,
+    };
+    const process = child_process.exec(
+      [getDefaultBuildifierExecutablePath()].concat(args).join(" "),
```

Comment:
If we ever use `--path` this will be an issue because of space.

Reply:
Added a comment.